plumbing/transport/common.go

```diff
@@ -88,42 +89,135 @@ type ReceivePackSession interface {
+	User() string
+	// Password returns the password or an empty string if none is given.
+	Password() string
+	// User returns the user or an empty string if none is given.
+	Host() string
+	// Port returns the port or 0 if there is no port or a default should be
+	// used.
+	Port() int
+	// Path returns the repository path.
+	Path() string
+	// String returns a string representation of the Git URL.
+	String() string
+}
 
 func NewEndpoint(endpoint string) (Endpoint, error) {
-	endpoint = transformSCPLikeIfNeeded(endpoint)
+	e := parseSCPLike(endpoint)
+	if e != nil {
+		return e, nil
+	}
 
 	u, err := url.Parse(endpoint)
 	if err != nil {
-		return Endpoint{}, plumbing.NewPermanentError(err)
+		return nil, plumbing.NewPermanentError(err)
 	}
 
 	if !u.IsAbs() {
-		return Endpoint{}, plumbing.NewPermanentError(fmt.Errorf(
+		return nil, plumbing.NewPermanentError(fmt.Errorf(
 			"invalid endpoint: %s", endpoint,
 		))
 	}
 
-	return Endpoint(*u), nil
+	return urlEndpoint{u}, nil
+}
+
+type urlEndpoint struct {
+	*url.URL
+}
+
+func (e urlEndpoint) Protocol() string { return e.URL.Scheme }
+func (e urlEndpoint) Host() string     { return e.URL.Hostname() }
+
+func (e urlEndpoint) User() string {
+	if e.URL.User == nil {
+		return ""
+	}
+
+	return e.URL.User.Username()
 }
 
-func (e *Endpoint) String() string {
-	u := url.URL(*e)
-	return u.String()
+func (e urlEndpoint) Password() string {
+	if e.URL.User == nil {
+		return ""
+	}
+
+	p, _ := e.URL.User.Password()
+	return p
 }
 
-func transformSCPLikeIfNeeded(endpoint string) string {
-	if !isSchemeRegExp.MatchString(endpoint) && scpLikeUrlRegExp.MatchString(endpoint) {
-		m := scpLikeUrlRegExp.FindStringSubmatch(endpoint)
-		return fmt.Sprintf("ssh://%s%s/%s", m[1], m[2], m[3])
+func (e urlEndpoint) Port() int {
+	p := e.URL.Port()
+	if p == "" {
+		return 0
 	}
 
-	return endpoint
+	i, err := strconv.Atoi(e.URL.Port())
+	if err != nil {
+		return 0
+	}
+
+	return i
+}
+
+func (e urlEndpoint) Path() string {
+	var res string = e.URL.Path
+	if e.URL.RawQuery != "" {
+		res += "?" + e.URL.RawQuery
+	}
+
+	if e.URL.Fragment != "" {
+		res += "#" + e.URL.Fragment
+	}
+
+	return res
+}
+
+type scpEndpoint struct {
+	user string
+	host string
+	path string
+}
+
+func (e *scpEndpoint) Protocol() string { return "ssh" }
+func (e *scpEndpoint) User() string     { return e.user }
+func (e *scpEndpoint) Password() string { return "" }
+func (e *scpEndpoint) Host() string     { return e.host }
+func (e *scpEndpoint) Port() int        { return 22 }
+func (e *scpEndpoint) Path() string     { return e.path }
+
+func (e *scpEndpoint) String() string {
+	var user string
+	if e.user != "" {
+		user = fmt.Sprintf("%s@", e.user)
+	}
+
+	return fmt.Sprintf("%s%s:%s", user, e.host, e.path)
+}
+
+var (
+	isSchemeRegExp   = regexp.MustCompile("^[^:]+://")
+	scpLikeUrlRegExp = regexp.MustCompile("^(?:(?P<user>[^@]+)@)?(?P<host>[^:]+):/?(?P<path>.+)$")
+)
+
+func parseSCPLike(endpoint string) Endpoint {
```

Comment:
Returning `(Endpoint, bool)` would be more idiomatic, so it can be used like:
```
if e, ok := parseSCPLike(s) ; ok {
    return e, nil
}
// try other formats...
```